### DISCUSSION TITLE
Error="function not implemented"

### DISCUSSION THREAD

**Discussion by polestar1988** [gramineproject/gramine#1735] [General]
@dimakuv 
I use Storm In my application in order to to store and retrieve  adapter information , Storm uses boltdb, which is a key-value store based on mmap, to manage the data.
I have created docker image using gsc and started the container inside the sgx.
when I want to add a datasource I get error `Error="function not implemented" `, while app is working without sgx perefctly.
the error occurs in line: 
`catalog, err := storm.Open(absCatalogPath)`
Is sgx or gramine support mmap files like storm? how could I solve the problem?

**Comment by dimakuv** [comment]
> when I want to add a datasource I get error Error="function not implemented" , while app is working without sgx perefctly.

Could you enable the Gramine log using `loader.log_level = "all"` and share the log here?

Without the detailed log, it's impossible to guess which exact function is not implemented.

**Reply by polestar1988** [reply]
@dimakuv I attached the logs of the reuslt docker logs command after adding loader.log_level = "all" to manifest:
[logs.txt](https://github.com/gramineproject/gramine/files/14066475/logs.txt)

**Reply by dimakuv** [reply]
You have errors like this line:
```
2024/01/26 15:09:53 ERROR adding new adapter to catalog failed, Error="storm open: /app/data/datasources.db, function not implemented"
```

The only suspicious thing from Gramine is this:
```
(libos_parser.c:1658:buf_write_all) [P1:T3:ConnectorApp] trace: ---- flock(10, LOCK_EX|LOCK_NB) ...
(libos_fcntl.c:312:libos_syscall_flock) [P1:T3:ConnectorApp] warning: The app tried to use flock, but it's turned off (sys.experimental__enable_flock = false)
(libos_parser.c:1658:buf_write_all) [P1:T3:ConnectorApp] trace: ---- return from flock(...) = -38
```

So please enable flock support in Gramine manifest by adding: `sys.experimental__enable_flock = true` and try again.

**Reply by polestar1988** [reply]
@dimakuv It worked Thanks so much for your help.
